merkletree.go

```diff
@@ -305,62 +346,66 @@ func (mt *MerkleTree) addNode(tx db.Tx, n *Node) (*Hash, error) {
 }
 
 // Get returns the value of the leaf for the given key
-func (mt *MerkleTree) Get(k *big.Int) (*big.Int, error) {
+func (mt *MerkleTree) Get(k *big.Int) (*big.Int, *big.Int, []*Hash, error) {
 	// verfy that k is valid and fit inside the Finite Field.
 	if !cryptoUtils.CheckBigIntInField(k) {
-		return nil, errors.New("Key not inside the Finite Field")
+		return nil, nil, nil, errors.New("Key not inside the Finite Field")
 	}
 
 	kHash := NewHashFromBigInt(k)
 	path := getPath(mt.maxLevels, kHash[:])
 
 	nextKey := mt.rootKey
+	var siblings []*Hash
 	for i := 0; i < mt.maxLevels; i++ {
 		n, err := mt.GetNode(nextKey)
 		if err != nil {
-			return nil, err
+			return nil, nil, nil, err
 		}
 		switch n.Type {
 		case NodeTypeEmpty:
-			return nil, ErrKeyNotFound
+			return big.NewInt(0), big.NewInt(0), siblings, ErrKeyNotFound
 		case NodeTypeLeaf:
 			if bytes.Equal(kHash[:], n.Entry[0][:]) {
-				return n.Entry[1].BigInt(), nil
+				return n.Entry[0].BigInt(), n.Entry[1].BigInt(), siblings, nil
 			} else {
-				return nil, ErrKeyNotFound
+				return n.Entry[0].BigInt(), n.Entry[1].BigInt(), siblings, ErrKeyNotFound
 			}
 		case NodeTypeMiddle:
 			if path[i] {
 				nextKey = n.ChildR
+				siblings = append(siblings, n.ChildL)
 			} else {
 				nextKey = n.ChildL
+				siblings = append(siblings, n.ChildR)
 			}
 		default:
-			return nil, ErrInvalidNodeFound
+			return nil, nil, nil, ErrInvalidNodeFound
 		}
 	}
 
-	return nil, ErrKeyNotFound
+	return nil, nil, nil, ErrKeyNotFound
 }
 
 // Update updates the value of a specified key in the MerkleTree, and updates
-// the path from the leaf to the Root with the new values.
-func (mt *MerkleTree) Update(k, v *big.Int) error {
+// the path from the leaf to the Root with the new values. Returns the
+// CircomProcessorProof.
+func (mt *MerkleTree) Update(k, v *big.Int) (*CircomProcessorProof, error) {
```

Comment:
I think it should be `UpdateAndGetCircomProof` just like `AddAndGetCircomProof`

Reply:
The thing is that the `AddAndGetCircomProof` has extra computation from `Add` pure method, while `Update` does not get much extra computation returning the CircomProof or not.
That's why there are 2 methods for Addition, one without returning the CircomProof for the cases where is not needed, and another one returning the CircomProof with the extra computation that requires. While for Update, it does not need much extra computation to generate the CircomProof, and that's why I've kept in a single method, where if not needed the CircomProof can be ignored without having much unused extra computation.